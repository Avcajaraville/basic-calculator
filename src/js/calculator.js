export class Calculator {
  constructor() {
    this.$display = null;
    this.$keyboard = null;
  }

  init() {
    this.resetCalculator();
    this.$display = document.getElementById('display');
    this.$keyboard = document.getElementById('keyboard');
    this.initKeyboard();
  }

  initKeyboard() {
    this.$keyboard.addEventListener('click', event => {
      const { target } = event;

      if (!target.matches('button')) {
        return;
      }

      const keyType = target.getAttribute('data-key');

      if (keyType === 'operator') {
        this.addOperator(target.value);
        this.updateDisplay();
        return;
      }

      if (keyType === 'decimal') {
        this.addDecimal();
        this.updateDisplay();
        return;
      }

      if (keyType === 'clear') {
        this.resetCalculator();
        this.updateDisplay();
        return;
      }

      this.updateDisplayValue(target.value);
      this.updateDisplay();
    });
  }

  updateDisplayValue(value) {
    if (this.waitingForSecondOperand) {
      this.displayValue = '0';
      this.waitingForSecondOperand = false;
    }
    this.displayValue =
      this.displayValue === '0' ? value : this.displayValue + value;
  }

  updateDisplay() {
    this.$display.innerText = this.displayValue.replace('.', this.dot);
  }

  addDecimal() {
    if (this.waitingForSecondOperand === true) {
      this.displayValue = `0${this.dot}`;
      this.waitingForSecondOperand = false;
      return;
    }

    if (!this.displayValue.includes(this.dot)) {
      this.displayValue += this.dot;
    }
  }

  addOperator(operator) {
    const inputValue = parseFloat(this.displayValue.replace(this.dot, '.'));
    if (this.operator === null && !isNaN(inputValue)) {
      this.firstOperand = inputValue;
    } else if (operator) {
      const result = this.calculate(
        this.firstOperand,
        inputValue,
        this.operator
      );
      this.displayValue = String(result);
      this.firstOperand = result;
    }
    this.operator = operator;
    this.waitingForSecondOperand = true;
  }

  calculate(firstOperand, secondOperand, operator) {
    if (operator === '+') {
      return firstOperand + secondOperand;
    } else if (operator === '-') {
      return firstOperand - secondOperand;
    } else if (operator === '*') {
      return firstOperand * secondOperand;
    } else if (operator === '/') {
      return firstOperand / secondOperand;
    }

    return secondOperand;
  }

  resetCalculator() {
    this.displayValue = '0';
    this.dot = ',';
    this.operator = null;
    this.firstOperand = null;
    this.waitingForSecondOperand = false;
  }
}
