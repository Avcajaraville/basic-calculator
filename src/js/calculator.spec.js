import fs from 'fs';
import path from 'path';

import { Calculator } from './calculator';

describe('init', () => {
  let calculator = null;
  let html = null;

  beforeAll(done => {
    // eslint-disable-next-line no-undef
    fs.readFile(
      path.resolve(__dirname, '../index.html'),
      'utf8',
      (err, htmlFile) => {
        html = htmlFile.toString();
        done();
      }
    );
  });

  beforeEach(() => {
    document.documentElement.innerHTML = html;
    calculator = new Calculator();
  });

  describe('#constructor', () => {
    it('should initialize references to needed DOM elements', () => {
      expect(calculator.$display).toBe(null);
      expect(calculator.$keyboard).toBe(null);
    });
  });

  describe('#init', () => {
    it('should call resetCalculator method', () => {
      const resetCalculatorFn = jest.spyOn(calculator, 'resetCalculator');

      calculator.init();

      expect(resetCalculatorFn).toHaveBeenCalled();
    });

    it('should get DOMElement references for display and keyboard DOMElements', () => {
      calculator.init();

      expect(calculator.$display).toMatchSnapshot();
      expect(calculator.$keyboard).toMatchSnapshot();
    });

    it('should call initKeyboard method', () => {
      const initKeyboardFn = jest.spyOn(calculator, 'initKeyboard');

      calculator.init();

      expect(initKeyboardFn).toHaveBeenCalled();
    });
  });

  describe('#initKeyboard', () => {
    beforeEach(() => {
      calculator.init();
    });

    it('should call addOperator & updateDisplay methods if button is type operator', () => {
      const addOperatorFn = jest.spyOn(calculator, 'addOperator');
      const updateDisplayFn = jest.spyOn(calculator, 'updateDisplay');
      const $button = document.querySelector('[data-key="operator"]');

      triggerClick($button);

      expect(addOperatorFn).toHaveBeenCalledWith($button.value);
      expect(updateDisplayFn).toHaveBeenCalled();
    });

    it('should call addDecimal & updateDisplay methods if button is type decimal', () => {
      const addDecimalFn = jest.spyOn(calculator, 'addDecimal');
      const updateDisplayFn = jest.spyOn(calculator, 'updateDisplay');
      const $button = document.querySelector('[data-key="decimal"]');

      triggerClick($button);

      expect(addDecimalFn).toHaveBeenCalled();
      expect(updateDisplayFn).toHaveBeenCalled();
    });

    it('should call resetCalculator & updateDisplay methods if button is type clear', () => {
      const resetCalculatorFn = jest.spyOn(calculator, 'resetCalculator');
      const updateDisplayFn = jest.spyOn(calculator, 'updateDisplay');
      const $button = document.querySelector('[data-key="clear"]');

      triggerClick($button);

      expect(resetCalculatorFn).toHaveBeenCalled();
      expect(updateDisplayFn).toHaveBeenCalled();
    });

    it('should call updateDisplayValue & updateDisplay methods if its a numeric button', () => {
      const updateDisplayValueFn = jest.spyOn(calculator, 'updateDisplayValue');
      const updateDisplayFn = jest.spyOn(calculator, 'updateDisplay');
      const $button = document.querySelector('[value="9"]');

      triggerClick($button);

      expect(updateDisplayValueFn).toHaveBeenCalledWith($button.value);
      expect(updateDisplayFn).toHaveBeenCalled();
    });

    it('should not call any of the above methods if clicked on an element that is not a button', () => {
      const addOperatorFn = jest.spyOn(calculator, 'addOperator');
      const updateDisplayFn = jest.spyOn(calculator, 'updateDisplay');
      const addDecimalFn = jest.spyOn(calculator, 'addDecimal');
      const resetCalculatorFn = jest.spyOn(calculator, 'resetCalculator');
      const updateDisplayValueFn = jest.spyOn(calculator, 'updateDisplayValue');
      const $element = document.querySelector('.keyboard-grid__li');

      triggerClick($element);

      expect(addOperatorFn).not.toHaveBeenCalled();
      expect(updateDisplayFn).not.toHaveBeenCalled();
      expect(addDecimalFn).not.toHaveBeenCalled();
      expect(resetCalculatorFn).not.toHaveBeenCalled();
      expect(updateDisplayValueFn).not.toHaveBeenCalled();
    });
  });

  describe('#updateDisplayValue', () => {
    beforeEach(() => {
      calculator.init();
    });

    describe('waitingForSecondOperand is true', () => {
      it('should set display value and waitingForSecondOperand to false if waitingForSecondOperand is true', () => {
        calculator.waitingForSecondOperand = true;
        const newValue = '30';

        calculator.updateDisplayValue(newValue);

        expect(calculator.displayValue).toBe(newValue);
        expect(calculator.waitingForSecondOperand).toBe(false);
      });
    });

    describe('waitingForSecondOperand is false', () => {
      it('should set new display value equal to input if previous displayValue is 0', () => {
        calculator.displayValue = '0';
        const newValue = '30';

        calculator.updateDisplayValue(newValue);

        expect(calculator.displayValue).toBe(newValue);
      });

      it('should append new display value if previous displayValue is not 0', () => {
        const previousValue = '3';
        const newValue = '4';
        calculator.displayValue = previousValue;

        calculator.updateDisplayValue(newValue);

        expect(calculator.displayValue).toBe(`${previousValue}${newValue}`);
      });
    });
  });

  describe('#updateDisplay', () => {
    beforeEach(() => {
      calculator.init();
    });

    it('should display current displayValue on display DOMElement with proper decimal separator', () => {
      const displayValue = '3.4';
      calculator.displayValue = displayValue;

      calculator.updateDisplay();

      // NOTE: I rather use toMatchSnapshot but innerText is not implemented in jsdom, which is used by jest
      // see: https://github.com/jsdom/jsdom/issues/1245
      expect(calculator.$display.innerText).toBe(
        displayValue.replace('.', ',')
      );
    });
  });

  describe('#addDecimal', () => {
    beforeEach(() => {
      calculator.init();
    });

    describe('waitingForSecondOperand is true', () => {
      it('should set displayValue to 0,', () => {
        calculator.waitingForSecondOperand = true;

        calculator.addDecimal();

        expect(calculator.displayValue).toBe(`0${calculator.dot}`);
        expect(calculator.waitingForSecondOperand).toBe(false);
      });
    });

    describe('waitingForSecondOperand is false', () => {
      it('should append decimal separator to current displayValue', () => {
        const displayValue = '3';
        calculator.displayValue = displayValue;

        calculator.addDecimal();

        expect(calculator.displayValue).toBe(
          `${displayValue}${calculator.dot}`
        );
      });

      it('should not append decimal separator if current displayValue already has one', () => {
        const displayValue = '3,';
        calculator.displayValue = displayValue;

        calculator.addDecimal();

        expect(calculator.displayValue).toBe(displayValue);
      });
    });
  });

  describe('#addOperator', () => {
    beforeEach(() => {
      calculator.init();
    });

    describe('operator is null and displayValue is a number (when hitting an operator button for first time)', () => {
      it('should save displayValue in firstOperand and set waitingForSecondOperand to true', () => {
        calculator.displayValue = '3';

        calculator.addOperator(null);

        expect(calculator.firstOperand).toBe(
          parseFloat(calculator.displayValue)
        );
        expect(calculator.waitingForSecondOperand).toBe(true);
      });
    });

    describe('operator is not null (when hitting an operator for second or more times)', () => {
      it('should call calculate method and store result in displayValue, update displayValue, save operator and set waitingForSecondOperand to true', () => {
        const firstOperand = 3;
        const result = 30;
        const operator = '+';
        calculator.operator = operator;
        calculator.displayValue = `${firstOperand}`;
        calculator.firstOperand = firstOperand;
        const calculateFn = jest
          .spyOn(calculator, 'calculate')
          .mockReturnValue(result);

        calculator.addOperator(operator);

        expect(calculateFn).toHaveBeenCalledWith(
          firstOperand,
          firstOperand,
          operator
        );
        expect(calculator.displayValue).toBe(`${result}`);
        expect(calculator.firstOperand).toBe(result);
        expect(calculator.operator).toBe(operator);
        expect(calculator.waitingForSecondOperand).toBe(true);
      });
    });

    describe('#addOperator', () => {
      beforeEach(() => {
        calculator.init();
      });

      it('should add numbers if operator is +', () => {
        const a = 3;
        const b = 2;

        expect(calculator.calculate(a, b, '+')).toBe(a + b);
      });

      it('should substract numbers if operator is -', () => {
        const a = 3;
        const b = 2;

        expect(calculator.calculate(a, b, '-')).toBe(a - b);
      });

      it('should multiply numbers if operator is *', () => {
        const a = 3;
        const b = 2;

        expect(calculator.calculate(a, b, '*')).toBe(a * b);
      });

      it('should divide numbers if operator is /', () => {
        const a = 6;
        const b = 2;

        expect(calculator.calculate(a, b, '/')).toBe(a / b);
      });

      it('should return second operator if no operator is given', () => {
        const a = 3;
        const b = 2;

        expect(calculator.calculate(a, b)).toBe(b);
      });
    });

    describe('#resetCalculator', () => {
      beforeEach(() => {
        calculator.init();
      });

      it('should reset calculator values', () => {
        calculator.resetCalculator();

        expect(calculator.displayValue).toBe('0');
        expect(calculator.dot).toBe(',');
        expect(calculator.operator).toBe(null);
        expect(calculator.firstOperand).toBe(null);
        expect(calculator.waitingForSecondOperand).toBe(false);
      });
    });
  });

  function triggerClick($element) {
    $element.dispatchEvent(new Event('click', { bubbles: true }));
  }
});
