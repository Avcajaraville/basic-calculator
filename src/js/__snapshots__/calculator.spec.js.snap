// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init #init should get DOMElement references for display and keyboard DOMElements 1`] = `
<p
  aria-live="polite"
  class="calculator-result"
  id="display"
  role="region"
>
  
          0
        
</p>
`;

exports[`init #init should get DOMElement references for display and keyboard DOMElements 2`] = `
<ul
  class="keyboard-grid"
  id="keyboard"
>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      data-key="operator"
      value="+"
    >
      +
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      data-key="operator"
      value="-"
    >
      -
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      data-key="operator"
      value="*"
    >
      ร
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      data-key="operator"
      value="/"
    >
      รท
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="7"
    >
      7
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="8"
    >
      8
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="9"
    >
      9
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="4"
    >
      4
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="5"
    >
      5
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="6"
    >
      6
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="1"
    >
      1
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="2"
    >
      2
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="3"
    >
      3
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      value="0"
    >
      0
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn"
      data-key="decimal"
    >
      ,
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li"
  >
    
            
    <button
      class="btn btn--clear"
      data-key="clear"
    >
      ac
    </button>
    
          
  </li>
  
          
  <li
    class="keyboard-grid__li keyboard-grid__li--equal"
  >
    
            
    <button
      class="btn btn--equal"
      data-key="operator"
      value="="
    >
      
              =
            
    </button>
    
          
  </li>
  
        
</ul>
`;
